File: app/src/main/java/workshop1024/com/xproject/model/news/source/local/NewsLocalDataSource.java
package workshop1024.com.xproject.model.news.source.local;

import java.util.ArrayList;
import java.util.List;

import workshop1024.com.xproject.model.filter.Filter;
import workshop1024.com.xproject.model.news.News;
import workshop1024.com.xproject.model.news.source.NewsDataSource;
import workshop1024.com.xproject.model.publisher.Publisher;
import workshop1024.com.xproject.utils.ExecutorUtils;

/**
 * 新闻本地数据源
 */
public class NewsLocalDataSource implements NewsDataSource {
    private static NewsDataSource INSTANCE;

    private NewsDao mNewsDao;
    private ExecutorUtils mExecutorUtils;

    private NewsLocalDataSource(NewsDao newsDao, ExecutorUtils executorUtils) {
        mNewsDao = newsDao;
        mExecutorUtils = executorUtils;
    }

    public static NewsDataSource getInstance(NewsDao newsDao, ExecutorUtils executorUtils) {
        if (INSTANCE == null) {
            INSTANCE = new NewsLocalDataSource(newsDao, executorUtils);
        }
        return INSTANCE;
    }

    @Override
    public void getNewses(final LoadNewsCallback loadNewsCallback) {
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                final List<News> newsList = mNewsDao.getNewses();
                mExecutorUtils.getMainThreadExecutor().execute(new Runnable() {
                    @Override
                    public void run() {
                        if (newsList.isEmpty()) {
                            loadNewsCallback.onNewsLoaded(newsList);
                        } else {
                            loadNewsCallback.onDataNotAvaiable();
                        }
                    }
                });
            }
        };

        mExecutorUtils.getDiskIOExecutor().execute(runnable);
    }

    @Override
    public void getNewsByPublishers(final List<Publisher> publisherList, final LoadNewsCallback loadNewsCallback) {
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                //遍历分别获取每个发布者的新闻信息
                List<News> newsList = new ArrayList<>();
                for (Publisher publisher : publisherList) {
                    List<News> subNewsList = mNewsDao.getNewsByPublisherId(publisher.getPublisherId());
                    newsList.addAll(subNewsList);
                }

                if (newsList.isEmpty()) {
                    loadNewsCallback.onDataNotAvaiable();
                } else {
                    loadNewsCallback.onNewsLoaded(newsList);
                }
            }
        };

        mExecutorUtils.getDiskIOExecutor().execute(runnable);
    }

    @Override
    public void getNewsByFilter(Filter filter, LoadNewsCallback loadNewsCallback) {

    }

    @Override
    public void saveNews(News news) {

    }

    @Override
    public void deleteAllNewses() {

    }
}

File: news/src/main/java/workshop1024/com/xproject/news/model/news/local/NewsLocalDataSource.kt
package workshop1024.com.xproject.news.model.news.local

import android.util.Log
import workshop1024.com.xproject.base.utils.ExecutorUtils
import workshop1024.com.xproject.news.model.news.News
import workshop1024.com.xproject.news.model.news.source.NewsDataSource

class NewsLocalDataSource private constructor(private val mNewsDao: NewsDao, private val mExecutorUtils: ExecutorUtils) : NewsDataSource {

    override fun getNewsesByTypeAndKey(searchType: String, searchKey: String, loadCallback: NewsDataSource.LoadCallback) {
        Log.i("XProject", "NewsLocalDataSource getNewsesByTypeAndKey, searchType = $searchType searchKey = $searchKey")
        val getNewsesByTypeAndKeyRunnable = Runnable {
            val newsList = mNewsDao.getNewsesByTypeAndKey(searchType, searchKey)
            mExecutorUtils.mMainThreadExecutor.execute(Runnable {
                if (!newsList.isEmpty()) {
                    (loadCallback as NewsDataSource.LoadCacheOrLocalNewsesCallback).onCachedOrLocalNewsLoaded(newsList)
                } else {
                    loadCallback.onDataNotAvaiable()
                }
            })
        }

        mExecutorUtils.mDiskIOExecutor.execute(getNewsesByTypeAndKeyRunnable)
    }

    override fun deleteAllNewsesByTypeAndKey(searchType: String, searchKey: String) {
        Log.i("XProject", "NewsLocalDataSource deleteAllNewsesByTypeAndKey, searchType = $searchType searchKey = $searchKey")
        val deleteAllNewsesByTypeAndKeyRunnable = Runnable {
            mNewsDao.deleteAllNewsesByTypeAndKey(searchType, searchKey)
        }

        mExecutorUtils.mDiskIOExecutor.execute(deleteAllNewsesByTypeAndKeyRunnable)
    }

    override fun markNewsesReadedByNewsId(searchType: String, searchKey: String, newsIdList: List<String>) {
        Log.i("XProject", "NewsLocalDataSource markNewsesReadedByNewsId, searchType = $searchType searchKey = $searchKey newsIdList = ${newsIdList.toString()}")
        val markNewsesReadedByNewsIdRunnable = Runnable {
            for (newsId in newsIdList) {
                mNewsDao.markNewsReadedByNewsId(newsId)
            }
        }

        mExecutorUtils.mDiskIOExecutor.execute(markNewsesReadedByNewsIdRunnable)
    }

    override fun addNews(news: News) {
        Log.i("XProject", "NewsLocalDataSource addNews, news = ${news.toString()}")
        val addNewsRunnable = Runnable {
            mNewsDao.addNews(news)
        }

        mExecutorUtils.mDiskIOExecutor.execute(addNewsRunnable)
    }

    override fun getIsRequestRemoteBySearchTypeAndKey(searchType: String, searchKey: String): Boolean {
        return false
    }

    override fun refreshBySearchTypeAndKey(searchType: String, searchKey: String, isRequestRemote: Boolean, isRequestCache: Boolean) {

    }


    companion object {
        private lateinit var INSTANCE: NewsLocalDataSource

        fun getInstance(newsDao: NewsDao, executorUtils: ExecutorUtils): NewsLocalDataSource {
            synchronized(NewsLocalDataSource::class.java) {
                if (!this::INSTANCE.isInitialized) {
                    INSTANCE = NewsLocalDataSource(newsDao, executorUtils)
                }
            }
            return INSTANCE
        }
    }
}